### PR TITLE
Caching: Fixes regression of the caching of null representations for missing dictionary items

### PR DESCRIPTION
**Pull request by AndyButland**
### Prerequisites

- [X] I have added steps to test this contribution in the description below

Addresses https://github.com/umbraco/Umbraco-CMS/issues/20336

### Description
In an earlier PR, we added an update to the dictionary repository such that it cached null values, and prevented multiple database hits for dictionary keys that did not exist: https://github.com/umbraco/Umbraco-CMS/pull/15576

We then resolved this issue raised that identified some issues resulting when using localised text in the backoffice (supported in Umbraco 13): https://github.com/umbraco/Umbraco-CMS/pull/19350

Unfortunately it seems this caused a regression on the original feature.

In this PR I've resolved the regression and tested that the backoffice translations work as expected.

I've also added integration tests to verify the cache behaviour.

### Testing

See replication steps described in https://github.com/umbraco/Umbraco-CMS/issues/20336

With the code in this PR in place, verify that you no longer see the database requests for missing dictionary items.

### Release

Once approved and merged for 13.12, this will need cherry-picking up for 16.4.

### CODE REVIEW

File: src/Umbraco.Core/Cache/AppCacheExtensions.cs [hunk 1]
```diff
@@ -43,24 +43,41 @@ public static void InsertCacheItem<T>(
...
+        if (result == null)
         {
             return default;
         }
 
+        // If we've retrieved the specific string that represents null in the cache, return it only if we are requesting it (via a typed request for a string).
+        // Otherwise consider it a null value.
+        if (result == (object)Cms.Core.Constants.Cache.NullRepresentationInCache)
+        {
+            return typeof(T) == typeof(string) ? (T)result : default;
```
[the remark this answers is not in the document]
Comment (AndyButland) [reply src/Umbraco.Core/Cache/AppCacheExtensions.cs]: Have applied.

File: tests/Umbraco.Tests.Integration/Umbraco.Infrastructure/Persistence/Repositories/DictionaryRepositoryTest.cs [hunk 2]
```diff
@@ -395,6 +407,116 @@ public void Can_Perform_GetDictionaryItemKeyMap_On_DictionaryRepository()
...
+        {
+            var dictionaryItem = repository.Get("Read More Updated");
+
+            Assert.IsNull(dictionaryItem);
+        }
+
+        // Modify the value directly in the database such that it now exists. As we don't have caching enabled on the repository we should get the new value.
+        using (var scope = ScopeProvider.CreateScope())
+        {
+            scope.Database.Execute("UPDATE cmsDictionary SET [key] = 'Read More Updated' WHERE [key] = 'Read More'");
```
[the remark this answers is not in the document]
Comment (AndyButland) [reply tests/Umbraco.Tests.Integration/Umbraco.Infrastructure/Persistence/Repositories/DictionaryRepositoryTest.cs:526]: Think this is OK in a test method.

File: src/Umbraco.Core/Cache/AppCacheExtensions.cs [hunk 3]
```diff
@@ -43,24 +43,41 @@ public static void InsertCacheItem<T>(
     public static T? GetCacheItem<T>(this IAppCache provider, string cacheKey)
     {
         var result = provider.Get(cacheKey);
-        if (IsRetrievedItemNull(result))
+        if (result == null)
         {
             return default;
         }
 
+        // If we've retrieved the specific string that represents null in the cache, return it only if we are requesting it (via a typed request for a string).
```
Comment (kjac) [inline src/Umbraco.Core/Cache/AppCacheExtensions.cs:51]: I'm not sure I understand this. If I read it correctly:

1. Given `T` is `string`, and
2. The `string` item identified by `cacheKey` does not exist, and
3. `null` values are being cached

...we return the value of `Cms.Core.Constants.Cache.NullRepresentationInCache` (`"*NULL*"`) to whomever requested the `string` for `cacheKey`. 

I'd expect a `null` value returned in all and every cases where no item could be identified `cacheKey` - no matter if we're caching `null` values or not.

Perhaps I'm missing something here?

Comment (AndyButland) [reply src/Umbraco.Core/Cache/AppCacheExtensions.cs:51]: I _think_ it's right - you will likely need to look at it in conjunction with `DefaultRepositoryCachePolicy.Get()`.

When caching null values, we can't actually store null - as there would be no way to distinguish between a "null as not found" or a "null cached value".  So instead we cache a special string that is representing null in the cache.

So in `DefaultRepositoryCachePolicy.Get()` we:
- Request the entity from the cache and return it if found.
- If we've got a null value back and the repository is caching null values, we have to distinguish between the "null as not found" and "null cached value" states. So we look for a string in the cache using the cache key. If found, we know we have a null cached value, and return null.  If not, means we have a cache miss, so we go to the database to get it it, cache it and return it.

So for `GetCacheItem`, if we are requesting anything and don't get a hit, we do return null (or strictly `default`)

But if we request an entity and we have a cache hit, we might have the entity or we might have the "null representation" string. If we've requested the "null representation" string and got it, we want to return the "null representation" string and not null (lines 71-74), so the logic described above in `DefaultRepositoryCachePolicy.Get()` can work.  If we've not requested the "null representation" string, we've requested an entity, want to do the `TryConvertTo` to return the entity from the retrieved object in the cache.

Don't mind admitting that took me 15 minutes or so to get straight in my head again!

Comment (kjac) [reply src/Umbraco.Core/Cache/AppCacheExtensions.cs:51]: Alright that's making more sense. I'll give it a spin 👍

**Review by kjac — APPROVED** [review]
@AndyButland I have added a bit more to the tests, to verify that cached null values are also flushed as expected.

If you think it makes sense, feel free to merge 👍

**Comment by AndyButland** [comment]
It does - thanks.  Will do (and I'll cherry-pick for the 13.11 release and 16).

**Comment by AndyButland** [comment]
Port to 16 is in https://github.com/umbraco/Umbraco-CMS/pull/20349 - I'll merge when the build checks are happy, as it's effectively a cherry-pick.

**Comment by umbracocommunity** [comment]
This pull request has been mentioned on **Umbraco community forum**. There might be relevant details there:

https://forum.umbraco.com/t/dictionary-cache-regression/6041/6

**Comment by umbracocommunity** [comment]
This pull request has been mentioned on **Umbraco community forum**. There might be relevant details there:

https://forum.umbraco.com/t/website-significantly-slower-since-upgrading-from-v13-to-v16/6049/9
